### PR TITLE
Added server

### PR DESCRIPTION
**Pull request by Paulls20**

### CODE REVIEW

File: client/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+find_package(Threads)
+
+add_executable(${PROJECT_NAME}
+    main.cpp
+    tcp_client.cpp)
+
+target_link_libraries(${PROJECT_NAME}
+    ${Boost_LIBRARIES})
+
+install(TARGETS ${PROJECT_NAME} DESTINATION ${CMAKE_SOURCE_DIR}/bin)
```
Comment (mbharatheesha) [inline client/CMakeLists.txt:17]: are install targets always necessary to be specified? Just out of curiosity.

File: client/main.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
+#include <boost/program_options.hpp>
+#include <iostream>
+#include "tcp_client.hpp"
+
+int main(int argc, const char *argv[])
+{
+    uint16_t port_num = 3333;
+    std::string ip_addr = "127.0.0.1";
```
Comment (mbharatheesha) [inline client/main.cpp]: having these parameters read from a config file makes it nicer I think.

File: client/main.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    namespace po = boost::program_options;
```
Comment (mbharatheesha) [inline client/main.cpp]: Is this common practice to use `po` in the context of Boost? To me, `po` is something that diminishes the readability of the code.

File: client/tcp_client.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
+#include "tcp_client.hpp"
+#include <future>
+#include <boost/asio/use_future.hpp>
+#include <array>
+#include <iostream>
+
+namespace fizyr
```
Comment (mbharatheesha) [inline client/tcp_client.cpp]: I think its better to not use fizyr yet. Perhaps some other name, like `abc_client` would be better.

File: client/tcp_client.cpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+{
+constexpr uint8_t BUFFER_SIZE = 128;
+
+tcp_client::tcp_client(const std::string& ip_addr, uint16_t port) :
+    ip_addr_(ip_addr), port_num_(port), work_(ios_), sock_(ios_)
+{
+    thread_ = std::make_unique<std::thread>([this](){ ios_.run(); });
```
Comment (mbharatheesha) [inline client/tcp_client.cpp]: May be a comment here as to it is unique because it is associated with this client?

File: client/tcp_client.hpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    /**
+     * \brief Send request to the server.
+     * \param request the message send to server.
+     */
+    void send_request(const std::string& request);
+
+private:
+    std::string ip_addr_;
+    uint16_t port_num_;
+    basio::io_service ios_;
```
Comment (mbharatheesha) [inline client/tcp_client.hpp]: Renaming these varables would help with readability..Like
- `io_service_`
- `tcp_socket_`
- `client_thread_`
- ...

File: server/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    main.cpp
+    tcp_server.cpp
+    acceptor.cpp
+    service.cpp
+    server_base.cpp)
+
+target_link_libraries(${PROJECT_NAME}
+    ${Boost_LIBRARIES})
+
+install(TARGETS ${PROJECT_NAME} DESTINATION ${CMAKE_SOURCE_DIR}/bin)
```
Comment (mbharatheesha) [inline server/CMakeLists.txt:23]: same remark as with the client.

File: server/acceptor.hpp [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+ */
+class acceptor
+{
+public:
+    /**
+     * \brief acceptor constructor
+     * \param ios instance of 'io_service' class is required by socket constructor.
+     * \param port_num port number on which it will listen for the incoming connection
+     */
+    acceptor(basio::io_service &ios, unsigned short port_num);
```
Comment (mbharatheesha) [inline server/acceptor.hpp:24]: iiuc rule of 5 correctly, shouldn't you also have the remaining 4 defined along with the constructor?

File: server/tcp_server.cpp [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+{
+    acc_->start();
+    BOOST_LOG_TRIVIAL(info) << "Server Started. Listening to port " << port_num_;
+    return true;  // TODO: check should be done.
+}
+
+bool tcp_server::do_close()
+{
+    acc_->stop();
+    return true; // TODO: check should be done.
```
Comment (mbharatheesha) [inline server/tcp_server.cpp]: there are a couple of TODOs here. If you already did them, you can get rid of these comments.

File: client/CMakeLists.txt (hunk 1 above)
Comment (Paulls20) [reply client/CMakeLists.txt:17]: Install to the parent bin directory to make life easier.

File: client/tcp_client.cpp (hunk 4 above)
Comment (Paulls20) [reply client/tcp_client.cpp]: I'll change

File: server/CMakeLists.txt (hunk 7 above)
Comment (Paulls20) [reply server/CMakeLists.txt:23]: same as above

File: client/main.cpp (hunk 2 above)
Comment (Paulls20) [reply client/main.cpp]: Done.

File: client/main.cpp (hunk 3 above)
Comment (Paulls20) [reply client/main.cpp]: Done.

File: client/tcp_client.cpp (hunk 4 above)
Comment (Paulls20) [reply client/tcp_client.cpp]: Changed the namespace to async.

File: client/tcp_client.cpp (hunk 5 above)
Comment (Paulls20) [reply client/tcp_client.cpp]: Added comment.

File: client/tcp_client.hpp (hunk 6 above)
Comment (Paulls20) [reply client/tcp_client.hpp]: Renamed ios_ to io_service_ and sock_ to tcp_socket_.

File: server/tcp_server.cpp (hunk 9 above)
Comment (Paulls20) [reply server/tcp_server.cpp]: Keeping the TODO as we discussed.

**Review by bajosoto — APPROVED** [review]

**Pull request merged**